File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/inbox/chat/ConversationRow.kt
@file:Suppress("MagicNumber", "PackageNaming", "LongMethod")

package app.pantopus.android.ui.screens.inbox.chat

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.pantopus.android.ui.screens.inbox.conversation.ai.ChatAiAvatar
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusIcon
import app.pantopus.android.ui.theme.PantopusIconImage
import app.pantopus.android.ui.theme.Radii
import app.pantopus.android.ui.theme.Spacing

@Composable
fun ConversationRow(
    content: ConversationRowContent,
    onTap: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Box(
        modifier =
            modifier
                .fillMaxWidth()
                .background(
                    if (content.pinned) PantopusColors.primary50.copy(alpha = 0.5f) else PantopusColors.appSurface,
                )
                .clickable(onClick = onTap)
                .padding(horizontal = Spacing.s4, vertical = Spacing.s4)
                .semantics(mergeDescendants = true) {
                    contentDescription = buildA11yLabel(content)
                }
                .testTag("conversationRow_${content.id}"),
    ) {
        if (content.pinned) {
            Box(
                modifier =
                    Modifier
                        .align(Alignment.CenterStart)
                        .offset(x = (-16).dp)
                        .width(3.dp)
                        .height(48.dp)
                        .background(PantopusColors.primary600),
            )
        }
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(Spacing.s3),
        ) {
            Avatar(content)
            Middle(content, modifier = Modifier.weight(1f))
            Trailing(content)
        }
        Box(
            modifier =
                Modifier
                    .align(Alignment.BottomCenter)
                    .fillMaxWidth()
                    .height(1.dp)
                    .background(PantopusColors.appBorderSubtle),
        )
    }
}

@Composable
private fun Avatar(content: ConversationRowContent) {
    when (val variant = content.variant) {
        ConversationRowVariant.AiAssistant -> ChatAiAvatar(size = 44.dp)
        ConversationRowVariant.Dm -> DmAvatar(initials = content.initials, verified = content.verified)
        is ConversationRowVariant.Group -> GroupAvatar(content.initials, variant)
    }
}

@Composable
private fun DmAvatar(
    initials: String,
    verified: Boolean,
) {
    Box(modifier = Modifier.size(52.dp), contentAlignment = Alignment.BottomEnd) {
        InitialsCircle(
            size = 52.dp,
            color = PantopusColors.personalBg,
            initials = initials,
            fg = PantopusColors.primary600,
        )
        if (verified) {
            Box(
                modifier =
                    Modifier
                        .size(16.dp)
                        .clip(CircleShape)
                        .background(PantopusColors.primary600)
                        .border(2.dp, PantopusColors.appSurface, CircleShape)
                        .offset(x = 1.dp, y = 1.dp)
                        .semantics { contentDescription = "Verified" },
                contentAlignment = Alignment.Center,
            ) {
                PantopusIconImage(
                    icon = PantopusIcon.Check,
                    contentDescription = null,
                    size = 9.dp,
                    tint = PantopusColors.appTextInverse,
                )
            }
        }
    }
}

@Composable
private fun GroupAvatar(
    primary: String,
    group: ConversationRowVariant.Group,
) {
    Box(modifier = Modifier.size(46.dp)) {
        // Back tile
        Box(modifier = Modifier.offset(x = (-3).dp, y = (-3).dp)) {
            InitialsCircle(size = 32.dp, color = PantopusColors.warning, initials = primary)
        }
        // Front tile (extras)
        val front = group.extraAvatars.firstOrNull() ?: "+${group.extraCount.coerceAtLeast(1)}"
        Box(modifier = Modifier.align(Alignment.BottomEnd).offset(x = 3.dp, y = 3.dp)) {
            InitialsCircle(size = 32.dp, color = PantopusColors.success, initials = front)
        }
    }
}

@Composable
private fun InitialsCircle(
    size: Dp,
    color: Color,
    initials: String,
    fg: Color = PantopusColors.appTextInverse,
) {
    Box(
        modifier =
            Modifier
                .size(size)
                .clip(CircleShape)
                .background(color)
                .border(2.dp, PantopusColors.appSurface, CircleShape),
        contentAlignment = Alignment.Center,
    ) {
        Text(
            text = initials,
            fontSize = (size.value * 0.34f).sp,
            fontWeight = FontWeight.Bold,
            color = fg,
        )
    }
}

@Composable
private fun Middle(
    content: ConversationRowContent,
    modifier: Modifier = Modifier,
) {
    val isAiRow = content.variant == ConversationRowVariant.AiAssistant
    Column(modifier = modifier, verticalArrangement = Arrangement.spacedBy(Spacing.s1)) {
        Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(6.dp)) {
            Text(
                text = content.displayName,
                fontSize = 16.sp,
                fontWeight = if (content.unread > 0 || isAiRow) FontWeight.Bold else FontWeight.Medium,
                // A15.3 swaps the AI row's business purple for primary blues.
                color = if (isAiRow) PantopusColors.primary700 else PantopusColors.appText,
                maxLines = 1,
            )
            if (isAiRow) {
                Text(
                    text = "AI",
                    fontSize = 10.sp,
                    fontWeight = FontWeight.Bold,
                    color = PantopusColors.primary700,
                    modifier =
                        Modifier
                            .clip(RoundedCornerShape(Radii.pill))
                            .background(PantopusColors.primary50)
                            .padding(horizontal = 6.dp, vertical = 1.dp),
                )
            }
            content.identityChip?.let { IdentityDisclosureChip(chip = it) }
        }
        Text(
            text = content.preview,
            fontSize = 14.sp,
            fontWeight = if (content.unread > 0) FontWeight.SemiBold else FontWeight.Normal,
            color =
                when {
                    isAiRow -> PantopusColors.primary600
                    content.unread > 0 -> PantopusColors.appTextStrong
                    else -> PantopusColors.appTextSecondary
                },
            maxLines = 1,
        )
        if (content.topics.isNotEmpty()) {
            TopicPillsRow(topics = content.topics)
        }
    }
}

/**
 * Up to two topic pills under the preview plus a "+N" overflow pill.
 * The Column's spacedBy already contributes 4dp; the extra 2dp top
 * padding lands the designed 6dp gap.
 */
@Composable
private fun TopicPillsRow(topics: List<ConversationRowTopic>) {
    Row(
        modifier = Modifier.padding(top = 2.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(Spacing.s1),
    ) {
        topics.take(2).forEach { topic ->
            TopicPill(title = topic.title, icon = rowTopicIcon(topic.topicType))
        }
        val overflow = topics.size - 2
        if (overflow > 0) {
            TopicPill(title = "+$overflow", icon = null)
        }
    }
}

@Composable
private fun TopicPill(
    title: String,
    icon: PantopusIcon?,
) {
    Row(
        modifier =
            Modifier
                .clip(RoundedCornerShape(Radii.pill))
                .background(PantopusColors.appSurfaceSunken)
                .padding(horizontal = Spacing.s2, vertical = 3.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(3.dp),
    ) {
        if (icon != null) {
            PantopusIconImage(
                icon = icon,
                contentDescription = null,
                size = 10.dp,
                tint = PantopusColors.appTextSecondary,
            )
        }
        Text(
            text = title,
            fontSize = 11.sp,
            color = PantopusColors.appTextSecondary,
            maxLines = 1,
        )
    }
}

/** Icon per topic type — mirrors the conversation screen's topic strip. */
private fun rowTopicIcon(type: String): PantopusIcon =
    when (type) {
        "task", "gig" -> PantopusIcon.Briefcase
        "listing" -> PantopusIcon.Tag
        "marketplace" -> PantopusIcon.ShoppingBag
        else -> PantopusIcon.MessageCircle
    }

@Composable
private fun Trailing(content: ConversationRowContent) {
    if (content.variant == ConversationRowVariant.AiAssistant) {
        PantopusIconImage(
            icon = PantopusIcon.ChevronRight,
            contentDescription = null,
            size = 18.dp,
            tint = PantopusColors.primary600,
        )
        return
    }
    Column(
        horizontalAlignment = Alignment.End,
        verticalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(Spacing.s1),
        ) {
            if (content.isMuted) {
                PantopusIconImage(
                    icon = PantopusIcon.BellOff,
                    contentDescription = "Muted",
                    size = 14.dp,
                    tint = PantopusColors.appTextMuted,
                )
            }
            Text(
                text = content.timeLabel,
                fontSize = 12.sp,
                fontWeight = if (content.unread > 0) FontWeight.SemiBold else FontWeight.Normal,
                color = if (content.unread > 0) PantopusColors.primary600 else PantopusColors.appTextMuted,
            )
        }
        if (content.unread > 0) {
            Box(
                modifier =
                    Modifier
                        .heightIn(min = 20.dp)
                        .widthIn(min = 20.dp)
                        .clip(RoundedCornerShape(Radii.pill))
                        .background(PantopusColors.primary600)
                        .padding(horizontal = 6.dp, vertical = 1.dp)
                        .semantics { contentDescription = "${content.unread} unread" },
                contentAlignment = Alignment.Center,
            ) {
                Text(
                    text = "${content.unread}",
                    fontSize = 12.sp,
                    fontWeight = FontWeight.Bold,
                    color = PantopusColors.appTextInverse,
                )
            }
        }
    }
}

@Composable
private fun IdentityDisclosureChip(chip: ConversationIdentityChip) {
    val fg: Color
    val bg: Color
    val icon: PantopusIcon
    when (chip) {
        ConversationIdentityChip.Business -> {
            fg = PantopusColors.business
            bg = PantopusColors.businessBg
            icon = PantopusIcon.ShoppingBag
        }
        ConversationIdentityChip.Home -> {
            fg = PantopusColors.home
            bg = PantopusColors.homeBg
            icon = PantopusIcon.Home
        }
    }
    Row(
        modifier =
            Modifier
                .clip(RoundedCornerShape(Radii.xs))
                .background(bg)
                .padding(horizontal = 6.dp, vertical = 1.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(3.dp),
    ) {
        PantopusIconImage(icon = icon, contentDescription = null, size = Radii.md, tint = fg)
        Text(
            text = chip.label.uppercase(),
            fontSize = 9.sp,
            fontWeight = FontWeight.Bold,
            color = fg,
        )
    }
}

private fun buildA11yLabel(content: ConversationRowContent): String {
    val parts = mutableListOf<String>()
    parts.add(content.displayName)
    content.identityChip?.label?.let { parts.add(it) }
    if (content.verified) parts.add("verified")
    parts.add(content.preview)
    if (content.unread > 0) parts.add("${content.unread} unread")
    return parts.joinToString(". ")
}

File: frontend/apps/ios/Pantopus/Features/Chat/ConversationRow.swift
//
//  ConversationRow.swift
//  Pantopus
//
//  Per-row view for the Chat List. Three avatar treatments (DM /
//  group / AI-assistant), identity chip slot, pinned rail + tint, and
//  unread bolding rules baked in.
//

import SwiftUI

/// Single row in the chat list.
public struct ConversationRow: View {
    private let content: ConversationRowContent
    private let onTap: @MainActor () -> Void

    public init(content: ConversationRowContent, onTap: @escaping @MainActor () -> Void) {
        self.content = content
        self.onTap = onTap
    }

    public var body: some View {
        Button(action: onTap) {
            HStack(alignment: .center, spacing: Spacing.s3) {
                avatar
                middle
                trailing
            }
            .padding(.horizontal, Spacing.s4)
            .padding(.vertical, Spacing.s4)
            .background(rowBackground)
            .overlay(alignment: .leading) {
                if content.pinned {
                    Rectangle()
                        .fill(Theme.Color.primary600)
                        .frame(width: 3)
                        .clipShape(UnevenRoundedRectangle(
                            topLeadingRadius: 0,
                            bottomLeadingRadius: 0,
                            bottomTrailingRadius: 2,
                            topTrailingRadius: 2
                        ))
                }
            }
            .overlay(alignment: .bottom) {
                Rectangle()
                    .fill(Theme.Color.appBorderSubtle)
                    .frame(height: 1)
            }
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .accessibilityElement(children: .combine)
        .accessibilityLabel(accessibilityLabel)
        .accessibilityAddTraits(.isButton)
        .accessibilityIdentifier("conversationRow_\(content.id)")
    }

    private var rowBackground: some View {
        Group {
            if content.pinned {
                Theme.Color.primary50.opacity(0.5)
            } else {
                Theme.Color.appSurface
            }
        }
    }

    @ViewBuilder
    private var avatar: some View {
        switch content.variant {
        case .dm:
            DMAvatarView(content: content)
        case .aiAssistant:
            ChatAIAvatar(size: 44)
        case let .group(extras, extraCount):
            GroupAvatarView(content: content, extras: extras, extraCount: extraCount)
        }
    }

    private var isAIRow: Bool {
        if case .aiAssistant = content.variant { return true }
        return false
    }

    private var middle: some View {
        VStack(alignment: .leading, spacing: Spacing.s1) {
            HStack(spacing: 6) {
                Text(content.displayName)
                    .font(.system(size: 16, weight: content.unread > 0 || isAIRow ? .bold : .medium))
                    .foregroundStyle(isAIRow ? Theme.Color.primary700 : Theme.Color.appText)
                    .lineLimit(1)
                if isAIRow {
                    // A15.3 `.ai-badge` — primary identity, not business purple.
                    Text("AI")
                        .font(.system(size: 10, weight: .bold))
                        .foregroundStyle(Theme.Color.primary700)
                        .padding(.horizontal, 6)
                        .padding(.vertical, 1)
                        .background(Theme.Color.primary50, in: Capsule())
                        .accessibilityHidden(true)
                }
                if let chip = content.identityChip {
                    IdentityDisclosureChip(chip: chip)
                }
                if content.pinned {
                    Icon(.star, size: 11, color: Theme.Color.appTextMuted)
                        .accessibilityHidden(true)
                }
                Spacer(minLength: Spacing.s0)
            }
            Text(content.preview)
                .font(.system(size: 14, weight: content.unread > 0 ? .semibold : .regular))
                .foregroundStyle(
                    isAIRow
                        ? Theme.Color.primary600
                        : (content.unread > 0 ? Theme.Color.appTextStrong : Theme.Color.appTextSecondary)
                )
                .lineLimit(1)
            if !content.topics.isEmpty {
                topicPills
                    .padding(.top, 2)
            }
        }
    }

    /// Topic pills under the preview — first two topics + a "+N"
    /// overflow pill when the conversation has more.
    private var topicPills: some View {
        HStack(spacing: 4) {
            ForEach(content.topics.prefix(2)) { topic in
                ConversationTopicPill(topic: topic)
            }
            if content.topics.count > 2 {
                Text("+\(content.topics.count - 2)")
                    .font(.system(size: 11))
                    .foregroundStyle(Theme.Color.appTextSecondary)
                    .padding(.horizontal, 8)
                    .padding(.vertical, 3)
                    .background(Theme.Color.appSurfaceSunken, in: Capsule())
                    .accessibilityLabel("\(content.topics.count - 2) more topics")
            }
        }
        .accessibilityIdentifier("conversationRow.topics_\(content.id)")
    }

    @ViewBuilder private var trailing: some View {
        if isAIRow {
            Icon(.chevronRight, size: 18, color: Theme.Color.primary600)
                .accessibilityHidden(true)
        } else {
            trailingDefault
        }
    }

    private var trailingDefault: some View {
        VStack(alignment: .trailing, spacing: 6) {
            HStack(spacing: 4) {
                if content.isMuted {
                    Icon(.bellOff, size: 14, color: Theme.Color.appTextMuted)
                        .accessibilityLabel("Muted")
                }
                Text(content.timeLabel)
                    .font(.system(size: 12, weight: content.unread > 0 ? .semibold : .regular))
                    .foregroundStyle(content.unread > 0 ? Theme.Color.primary600 : Theme.Color.appTextMuted)
            }
            if content.unread > 0 {
                Text("\(content.unread)")
                    .font(.system(size: 12, weight: .bold))
                    .foregroundStyle(Theme.Color.appTextInverse)
                    .padding(.horizontal, 6)
                    .frame(minWidth: 20, minHeight: 20)
                    .background(Theme.Color.primary600)
                    .clipShape(Capsule())
                    .accessibilityLabel("\(content.unread) unread")
            }
        }
        .frame(minWidth: 32, alignment: .trailing)
    }

    private var accessibilityLabel: String {
        var parts: [String] = [content.displayName]
        if let chip = content.identityChip { parts.append(chip.label) }
        if content.verified { parts.append("verified") }
        parts.append(content.preview)
        if content.unread > 0 { parts.append("\(content.unread) unread") }
        return parts.joined(separator: ". ")
    }
}

// MARK: - Avatar variants

private struct DMAvatarView: View {
    let content: ConversationRowContent

    var body: some View {
        ZStack(alignment: .bottomTrailing) {
            initialsCircle(size: 52, color: Theme.Color.personalBg, initials: content.initials, fg: Theme.Color.primary600)
            if content.verified {
                Icon(.check, size: 9, strokeWidth: 3.5, color: Theme.Color.appTextInverse)
                    .frame(width: 16, height: 16)
                    .background(Theme.Color.primary600)
                    .clipShape(Circle())
                    .overlay(Circle().stroke(Theme.Color.appSurface, lineWidth: 2))
                    .offset(x: 1, y: 1)
                    .accessibilityLabel("Verified")
            }
        }
        .frame(width: 52, height: 52)
    }
}

private struct GroupAvatarView: View {
    let content: ConversationRowContent
    let extras: [String]
    let extraCount: Int

    var body: some View {
        ZStack {
            // Back tile
            initialsCircle(size: 32, color: Theme.Color.warning, initials: content.initials)
                .overlay(Circle().stroke(Theme.Color.appSurface, lineWidth: 2))
                .offset(x: -6, y: -6)
            // Front tile
            initialsCircle(
                size: 32,
                color: Theme.Color.success,
                initials: extras.first ?? "+\(extraCount > 0 ? extraCount : 1)"
            )
            .overlay(Circle().stroke(Theme.Color.appSurface, lineWidth: 2))
            .offset(x: 6, y: 6)
        }
        .frame(width: 44, height: 44)
        .accessibilityLabel("Group")
    }
}

private func initialsCircle(
    size: CGFloat,
    color: Color,
    initials: String,
    fg: Color = Theme.Color.appTextInverse
) -> some View {
    ZStack {
        Circle().fill(color)
        Text(initials)
            .font(.system(size: size * 0.34, weight: .bold))
            .foregroundStyle(fg)
    }
    .frame(width: size, height: size)
}

// MARK: - Topic pill

/// One topic pill under the preview line. The icon follows the topic
/// type: `task`/`gig` → briefcase, `listing`/`marketplace` → tag.
private struct ConversationTopicPill: View {
    let topic: ConversationRowTopic

    var body: some View {
        HStack(spacing: 3) {
            Icon(icon, size: 10, color: Theme.Color.appTextSecondary)
            Text(topic.title)
                .font(.system(size: 11))
                .foregroundStyle(Theme.Color.appTextSecondary)
                .lineLimit(1)
        }
        .padding(.horizontal, 8)
        .padding(.vertical, 3)
        .background(Theme.Color.appSurfaceSunken, in: Capsule())
    }

    private var icon: PantopusIcon {
        switch topic.topicType {
        case "task", "gig": .briefcase
        case "listing", "marketplace": .tag
        default: .messageCircle
        }
    }
}

// MARK: - Identity disclosure chip

private struct IdentityDisclosureChip: View {
    let chip: ConversationIdentityChip

    var body: some View {
        HStack(spacing: 3) {
            Icon(icon, size: 8, strokeWidth: 2.6, color: foreground)
            Text(chip.label.uppercased())
                .font(.system(size: 9, weight: .bold))
                .tracking(0.5)
                .foregroundStyle(foreground)
        }
        .padding(.horizontal, 6)
        .padding(.vertical, 1)
        .background(background)
        .clipShape(RoundedRectangle(cornerRadius: Radii.xs, style: .continuous))
    }

    private var icon: PantopusIcon {
        switch chip {
        case .business: .shoppingBag
        case .home: .home
        }
    }

    private var foreground: Color {
        switch chip {
        case .business: Theme.Color.business
        case .home: Theme.Color.home
        }
    }

    private var background: Color {
        switch chip {
        case .business: Theme.Color.businessBg
        case .home: Theme.Color.homeBg
        }
    }
}

// MARK: - Swipe actions

/// Reveals mute + hide actions when the user swipes a conversation row left.
public struct SwipeableConversationRow: View {
    private let content: ConversationRowContent
    private let onTap: @MainActor () -> Void
    private let onMute: @MainActor () -> Void
    private let onHide: @MainActor () -> Void

    private let revealWidth: CGFloat = 140
    @State private var offset: CGFloat = 0
    @GestureState private var dragging: CGFloat = 0

    public init(
        content: ConversationRowContent,
        onTap: @escaping @MainActor () -> Void,
        onMute: @escaping @MainActor () -> Void,
        onHide: @escaping @MainActor () -> Void
    ) {
        self.content = content
        self.onTap = onTap
        self.onMute = onMute
        self.onHide = onHide
    }

    public var body: some View {
        ZStack(alignment: .trailing) {
            swipeActions
            ConversationRow(content: content, onTap: onTap)
                .offset(x: clampedOffset)
                .highPriorityGesture(swipeGesture)
        }
        .accessibilityIdentifier("swipeableConversationRow_\(content.id)")
    }

    private var clampedOffset: CGFloat {
        min(0, max(-revealWidth, offset + dragging))
    }

    private var swipeActions: some View {
        HStack(spacing: 0) {
            swipeButton(
                icon: content.isMuted ? .bell : .bellOff,
                label: content.isMuted ? "Unmute" : "Mute",
                tint: Color(red: 0.42, green: 0.45, blue: 0.50),
                id: "conversationRow.swipeMute_\(content.id)"
            ) {
                reset()
                onMute()
            }
            swipeButton(
                icon: .archive,
                label: "Hide",
                tint: Color(red: 0.22, green: 0.25, blue: 0.32),
                id: "conversationRow.swipeHide_\(content.id)"
            ) {
                reset()
                onHide()
            }
        }
        .frame(width: revealWidth)
        .opacity(clampedOffset < -2 ? 1 : 0)
    }

    private func swipeButton(
        icon: PantopusIcon,
        label: String,
        tint: Color,
        id: String,
        action: @escaping @MainActor () -> Void
    ) -> some View {
        Button(action: action) {
            VStack(spacing: 3) {
                Icon(icon, size: 17, color: Theme.Color.appTextInverse)
                Text(label)
                    .font(.system(size: 11, weight: .semibold))
                    .foregroundStyle(Theme.Color.appTextInverse)
            }
            .frame(maxWidth: .infinity, maxHeight: .infinity)
            .background(tint)
        }
        .buttonStyle(.plain)
        .accessibilityIdentifier(id)
    }

    private var swipeGesture: some Gesture {
        DragGesture(minimumDistance: 14)
            .updating($dragging) { value, state, _ in
                if abs(value.translation.width) > abs(value.translation.height) {
                    state = value.translation.width
                }
            }
            .onEnded { value in
                let projected = offset + value.translation.width
                withAnimation(.interpolatingSpring(stiffness: 320, damping: 30)) {
                    offset = projected < -revealWidth / 2 ? -revealWidth : 0
                }
            }
    }

    private func reset() {
        withAnimation(.interpolatingSpring(stiffness: 320, damping: 30)) {
            offset = 0
        }
    }
}
